### PR TITLE
Add score-based diffusion model for STM downstream resampler and Stickman production target

### PR DESCRIPTION
**Pull request by YongyiBWu**
STM downstream simulations have suffered from low statistics and high computational expense. Here a score-based diffusion model is added to resample the particles at VD116 for reuse. 
 
Inconel 718 Stickman muon production target geometry is implemented from the production blueprints (as built). Support wheel implementation is corrected according to CRR drawings in docdb-30814. The geometry has been discussed with and blessed by Kevin and Michael. Overlap checks were performed and passed.

### CODE REVIEW

File: STMMC/src/VDResamplerConfigure_module.cc [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+      // Get the associated particle
+      const SimParticle& particle = SimParticles.at(step.trackId());
+      pdgId = particle.pdgId();
+
+      virtualdetectorId = step.virtualDetectorId();
+      pz = step.momentum().z();
+      if (virtualdetectorId != VirtualDetectorID || pz <= 0)
+      {
+        mf::LogWarning("VDResamplerConfigure") << "Thrown event\n"
+                                               << "PDG ID = " << pdgId << ", VDID = " << virtualdetectorId << ", z = " << step.position().z() << ", pz = " << pz;
```
[the remark this answers is not in the document]
Comment (YongyiBWu) [reply STMMC/src/VDResamplerConfigure_module.cc]: Addressed in a new commit; a new counter was added to tally these events and report at the summary.

File: STMMC/src/VDResamplerConfigure_module.cc [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    };
+    return;
+  };
+
+  void VDResamplerConfigure::endJob() {
+    mf::LogInfo log("Virtual detector tree summary");
+    log << "========= Data summary =========\n";
+    for (auto part : pdgIds)
+      log << "PDGID " << part.first << ": " << part.second << "\n";
+    log << "================================\n";
```
[the remark this answers is not in the document]
Comment (YongyiBWu) [reply STMMC/src/VDResamplerConfigure_module.cc]: Addressed in a new commit

File: STMMC/src/VDResamplerTrain_module.cc [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+#include "Offline/GlobalConstantsService/inc/GlobalConstantsHandle.hh"
+#include "Offline/GlobalConstantsService/inc/ParticleDataList.hh"
+#include "Offline/MCDataProducts/inc/SimParticle.hh"
+#include "Offline/MCDataProducts/inc/StepPointMC.hh"
+#include "Offline/SeedService/inc/SeedService.hh"
+
+// ROOT includes
+#include "art_root_io/TFileService.h"
+#include "TTree.h"
+
```
[the remark this answers is not in the document]
Comment (YongyiBWu) [reply STMMC/src/VDResamplerTrain_module.cc]: Addressed in a new commit

File: STMMC/src/VDResamplerTrain_module.cc [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+          conf().SBDMbetaMax(),
+          conf().SBDMcosineOffset(),
+          conf().SBDMbatchSize(),
+          conf().SBDMgradientClip(),
+          conf().SBDMlearningRate(),
+          conf().SBDMdiffusionSteps()
+      );
+
+      stage1TrainingData.reserve(1000000);
+      stage2TrainingData.reserve(1000000);
```
[the remark this answers is not in the document]
Comment (YongyiBWu) [reply STMMC/src/VDResamplerTrain_module.cc]: Addressed in a new commit. By default the training size is now passed through from the configurer, memory allocation is based on the number of training size now.

File: STMMC/src/VDResamplerTrain_module.cc [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    // Loop over all VD hits
+    for (const StepPointMC& step : StepPointMCs) {
+      // Get the associated particle
+      const SimParticle& particle = SimParticles.at(step.trackId());
+
+      // Extract the parameters
+      stepPdgId = particle.pdgId();
+      virtualdetectorId = step.virtualDetectorId();
+      time = step.time();
+      x = step.position().x(); // Shift the x coordinate to be relative to the centre of beamline
```
[the remark this answers is not in the document]
Comment (YongyiBWu) [reply STMMC/src/VDResamplerTrain_module.cc]: Addressed in a new commit

**Comment by oksuzian** [comment]
I'll review this pull request by examining its details and changes.Let me fetch the changed files and the existing review comments.Let me fetch the actual changed files from the PR's head branch to review the code:Now let me also check the existing review comments and the CI status:Here is my review of [PR #1794](https://github.com/Mu2e/Offline/pull/1794). This is a large PR (+4416/−11 across 20 files) with two distinct features: (1) a score-based diffusion model for STM downstream resampling, and (2) the Stickman production target geometry. I've identified several issues organized by severity.

---

## 🔴 Critical Issues

### 1. `RandFlat` / `RandGaussQ` stored by value instead of by reference — copies underlying distributions

```cpp name=MachineLearningTools/inc/ScoreBasedDiffusionModel.hh url=https://github.com/Mu2e/Offline/blob/9208a0e44239239b549d9d36f1c320c20fbd12a3/MachineLearningTools/inc/ScoreBasedDiffusionModel.hh#L283-L284
CLHEP::RandFlat   randFlat_;      // Used for uniform sampling (e.g., batch selection)
CLHEP::RandGaussQ randGaussQ_;    // Used for Gaussian noise in diffusion process
```

The constructor accepts `CLHEP::RandFlat&` and `CLHEP::RandGaussQ&` by reference, but the members `randFlat_` and `randGaussQ_` are stored **by value**, which means they are **copy-constructed** from the references. `CLHEP::RandFlat` and `RandGaussQ` hold a pointer to their underlying engine. Copying these objects will either share the engine in an unintended way or trigger undefined behavior depending on the CLHEP version. This is especially dangerous in `loadModel()` where the returned `ScoreBasedDiffusionModel` object is moved/copied. These should be stored as references or pointers:

```cpp
CLHEP::RandFlat&   randFlat_;
CLHEP::RandGaussQ& randGaussQ_;
```

### 2. `lastEnum` comment is inconsistent with actual value

```cpp name=MCDataProducts/inc/GenId.hh url=https://github.com/Mu2e/Offline/blob/9208a0e44239239b549d9d36f1c320c20fbd12a3/MCDataProducts/inc/GenId.hh#L47-L48
      gammaPairProduction, antiproton, Mu2eXGenTool, STMDownStreamGenTool,//56
      lastEnum //56
```

`STMDownStreamGenTool` is at index 56, so `lastEnum` should be 57, but the comment says `//56`. While the compiler will assign the correct value, the misleading comment could confuse developers validating enum ranges.

### 3. Default geometry switched from Hayman to Stickman for all users

```cpp name=Mu2eG4/geom/geom_run1_a.txt url=https://github.com/Mu2e/Offline/blob/9208a0e44239239b549d9d36f1c320c20fbd12a3/Mu2eG4/geom/geom_run1_a.txt#L69-L70
// #include "Offline/Mu2eG4/geom/ProductionTarget_Hayman_v2_2.txt"
#include "Offline/Mu2eG4/geom/ProductionTarget_Stickman_v1_0.txt"
```

This changes the **default production target for all `geom_run1_a` users** from Hayman to Stickman. If Stickman is not yet validated for all workflows or this is not the intended default, this could break existing simulations across the collaboration. Consider making this opt-in via a separate geometry file or keeping Hayman as default with a new `geom_run1_stickman.txt`.

---

## 🟡 Significant Issues

### 4. Weight initialization uses wrong scale

```cpp name=MachineLearningTools/src/ScoreBasedDiffusionModel.cc url=https://github.com/Mu2e/Offline/blob/9208a0e44239239b549d9d36f1c320c20fbd12a3/MachineLearningTools/src/ScoreBasedDiffusionModel.cc#L100-L101
const double weightInitScale = 0.02;
// const double weightInitScale = std::sqrt(2.0 / in); // He initialization for ReLU activations
```

The code comments say it generates `N(0, weightInitScale)` but then calls:
```cpp
layer.W[i][j] = weightInitScale * randGaussQ_.fire(); 
```
`RandGaussQ::fire()` returns `N(0,1)`, so the actual standard deviation is `weightInitScale = 0.02`. The commented-out He initialization `sqrt(2/in)` would be more appropriate for SiLU (similar to ReLU). The fixed 0.02 may lead to vanishing gradients in deeper networks.

### 5. Gradient accumulation without batch normalization

In the training loop, gradients are accumulated across `batchSize_` samples but are never **divided by the batch size** before the optimizer step. This means the effective learning rate scales with batch size, which could cause training instability if the batch size changes:

```cpp name=MachineLearningTools/src/ScoreBasedDiffusionModel.cc url=https://github.com/Mu2e/Offline/blob/9208a0e44239239b549d9d36f1c320c20fbd12a3/MachineLearningTools/src/ScoreBasedDiffusionModel.cc#L327-L335
batchCounter++;
if(batchCounter == batchSize_)
{
    clipGradients(gradientClipThreshold_);
    if (optimizerType_ == OptimizerType::ADAM) {
        adamUpdate(learningRate_);
    } else {
        updateWeights(learningRate_);
    }
```

Gradients should be normalized: `layer.gradW[i][j] /= batchSize_` before the update, or the learning rate should be documented as being batch-size dependent.

### 6. `HepLorentzVector` constructor argument order

```cpp name=STMMC/src/VDResamplerGenerateFromModel_module.cc url=https://github.com/Mu2e/Offline/blob/9208a0e44239239b549d9d36f1c320c20fbd12a3/STMMC/src/VDResamplerGenerateFromModel_module.cc#L267
const CLHEP::HepLorentzVector fourMomParticle(eTotal, momParticle);
```

`CLHEP::HepLorentzVector(double, Hep3Vector)` constructs `(t, x, y, z)` — i.e. the scalar first is the **time/energy component** and the 3-vector is the **spatial/momentum** component. So `HepLorentzVector(eTotal, momParticle)` sets `E = eTotal` and `p = momParticle`. This happens to be correct for the `(E, px, py, pz)` convention, but the same pattern in `VDResamplerGenerateMix_module.cc` should be verified. The more explicit constructor `HepLorentzVector(px, py, pz, E)` would avoid ambiguity.

### 7. Hardcoded detector constants duplicated across multiple modules

The transform constants `x0_ = -3904.0`, `y0_ = 0.0`, `t0_ = 1700.0`, `tScale_ = 1.0`, `p0_ = 1.0` are **hardcoded identically** in three separate files:
- `VDResamplerTrain_module.cc`
- `VDResamplerGenerateFromModel_module.cc`
- `VDResamplerGenerateMix_module.cc`

If any of these changes, all three must be updated in lockstep. These should be defined once in a shared header or read from FHiCL configuration.

---

## 🟢 Minor / Style Issues

### 8. `ProductionTargetMaker.cc` error message references wrong config key

```cpp name=GeometryService/src/ProductionTargetMaker.cc url=https://github.com/Mu2e/Offline/blob/9208a0e44239239b549d9d36f1c320c20fbd12a3/GeometryService/src/ProductionTargetMaker.cc#L24
{throw cet::exception("GEOM") << " illegal production target version specified = " << c.getInt("targetPS_version")  << std::endl;}
```

The error message prints `targetPS_version` (an int), but the preceding logic branches on `c.getString("targetPS_model")` (a string). The error should print `c.getString("targetPS_model")` instead.

### 9. Missing `#include <numeric>` dependency not needed in the header

```cpp name=ProductionTargetGeom/inc/ProductionTarget.hh url=https://github.com/Mu2e/Offline/blob/9208a0e44239239b549d9d36f1c320c20fbd12a3/ProductionTargetGeom/inc/ProductionTarget.hh#L10
#include <numeric>
```

`<numeric>` is added to the header but `std::accumulate` is only used in the `.cc` file. The include should be moved to `ProductionTarget.cc` to minimize header pollution.

### 10. Inconsistent coding style in `constructTargetPS.cc`

The new Stickman section (1000+ lines) uses `auto&` captures in lambdas, `reg.add()` calls, and range-based patterns while surrounding Hayman code uses different conventions. While functional, the formatting inconsistencies (indentation levels, brace placement, spacing) make the file harder to read. Consider following the existing style in the file.

### 11. Several `assert()` statements that should be exceptions

The diffusion model code uses `assert()` liberally (e.g., checking network size, input dimensions). In a release build with `NDEBUG`, these checks evaporate. For production code in a physics framework, `cet::exception` or at minimum conditional throws would be safer.

---

## Summary

| Severity | Count | Key Items |
|---|---|---|
| 🔴 Critical | 3 | RNG member copy semantics, default geometry swap, enum comment mismatch |
| 🟡 Significant | 4 | Gradient normalization, hardcoded constants, weight init, 4-vector constructor |
| 🟢 Minor | 4 | Error message, include placement, style consistency, assert vs exception |

The ML framework and Stickman geometry are well-documented and thoughtfully structured overall. The most impactful issues to address before merge are the **RNG storage semantics** (issue #1) and the **default geometry swap** (issue #3), as they affect correctness and all downstream users respectively.

**Comment by YongyiBWu** [comment]
Responses to the AI review:
1. Fixed as suggested.
2. Fixed as suggested.
3. Stickman is the actual as-built target. This should become the default geometry. The change is intentional. Yuri, if you prefer to have a geom_run1_a_stickman.txt and selects it in geom_common.txt, let me know.
4. Addressed. Using a capped reduced He initialization now (if reduced He = sqrt(2/size)/2 is smaller than 0.3, use reduced He, else use 0.3) to be scalable with input size.
5. Addressed. Good point. Added gradient averaging before gradient clipping and weight/bias update.
6. This is a non-issue. In both cases the Lorentz vector is constructed by first adding the momentum components to a Hep3Vector and then using CLHEP::HepLorentzVector(double, Hep3Vector) constructor. This explicitly gives the momentum for easier review.
7. Addressed. Added a header file in STMMC/inc that is now the single point of control.
8. Fixed as suggested.
9. Moved inclusion to ProductionTarget.cc as suggested.
10. This is an intended design improvement. The target plate construction is now modularized, the plates are added to registry and reused whenever possible, so they don't need to be reconstructed over and over again. Meanwhile, this method allows maximal flexibility leaving space for constructing place for different-sized target plates. The Hayman style only reuses simple geometry objects like Tubs and Boxes, following its style makes the construction more complex, so I chose format complexity over geometry construction complexity. 
11. Fixed as suggested.

**Comment by oksuzian** [comment]
"Yuri, if you prefer to have a geom_run1_a_stickman.txt and selects it in geom_common.txt, let me know." - yes

**Comment by oksuzian** [comment]
@FNALbuild run build test

**Comment by YongyiBWu** [comment]
Per request this PR was separated into two PRs: PR #1797 and PR #1798. This PR is now closed.
